plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -627,92 +663,96 @@ type ServiceAffinity struct {
 	labels        []string
 }
 
-func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) algorithm.FitPredicate {
+func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister algorithm.ServiceLister, nodeInfo NodeInfo, labels []string) (algorithm.FitPredicate, func(pm *predicateMetadata)) {
 	affinity := &ServiceAffinity{
 		podLister:     podLister,
 		serviceLister: serviceLister,
 		nodeInfo:      nodeInfo,
 		labels:        labels,
 	}
-	return affinity.CheckServiceAffinity
+
+	return affinity.CheckServiceAffinity, func(pm *predicateMetadata) {
+		var err error
+		// We can't know at the start what labels we are selecting against.
+		pm.servicePods[""], err = podLister.List(createSelectorFromLabels(nil))
```

Comment:
OK - sorry for not being clear.
What I meant is that once you create "RegisterPredicateMetadataProducerFactory" (that would use exactly the same mechanism as "RegisterPriorityMetadataProducerFactory", you would be able to write somethig like this (in default.go):
```
factory.RegisterPredicateMetadataProducerFactory(
  func(args factory.PluginFactoryArgs) algorithm.Metadataproducer {
    return predicates.NewPredicateMetadataFactory(args.PodLister)
  }
}
```

and define the factory like:
```
type PredicateMetadataFactory struct {
   podLister algorithm.PodLister
}

func NewPredicateMetadataFactory(podLister algorithm.PodLister) algorithm.MetadataProducer {
  factory := &PredicateMetadataFactory{podLister}
  return factory.GetMetadata
}

func (factory *PredicateMetadataFactory) Getmetadata(pod *api.Pod) PredicateMetadata {
  // compute the metadata for the given pod
  // but you already have access to "podLister" here (which you basically need)
}
```

Does it make sense now?